grails-app/controllers/org/pih/warehouse/product/ProductController.groovy

```diff
@@ -1117,23 +1117,12 @@ class ProductController {
         if (!enabled) {
             throw new IllegalArgumentException("Merge products feature is not enabled")
         }
-        Product primaryProduct = Product.get(params.primaryProduct)
-        Product obsoleteProduct = Product.get(params.obsoleteProduct)
 
-        if (!primaryProduct) {
-            throw new IllegalArgumentException("No Product found with ID ${params.primaryProduct}")
-        }
-        if (!obsoleteProduct) {
-            throw new IllegalArgumentException("No Product found with ID ${params.obsoleteProduct}")
-        }
-
-        if (primaryProduct == obsoleteProduct) {
-            throw new IllegalArgumentException("Cannot merge the product with itself")
-        }
+        productMergeService.validateProducts(params.primaryProduct, params.obsoleteProduct)
 
-        productMergeService.mergeProduct(primaryProduct, obsoleteProduct)
+        productMergeService.mergeProduct(params.primaryProduct, params.obsoleteProduct)
 
-        flash.message = "${obsoleteProduct.productCode} Product merge to ${primaryProduct.productCode} has succeeded. " +
+        flash.message = "${params.obsoleteProduct} Product merge to ${params.primaryProduct} has succeeded. " +
```

Comment:
Probably needs to be localized. We should use the same pattern as with saved, created, deleted, updated, etc just in case we eventually want to merge other domain objects.
```
default.created.message={0} {1} created
```
So ...

```
default.merged.message = "{0} {1} was successfully merged into {0} {2}"
```
and then flash.message would look like this
```
flash.message=${g:message(code: "default.merged.label" args="[g.message(code: "product.label"), params.obsoleteProduct, params.primaryProduct]")}"
```

Reply:
done